RFC: Implement @__LINE__ macro

ihnorton opened the pull request:
(upated) Implements `__LINE__` and `__FILE__` magic macros returning the line number (integer) and filename (as string) for a given call location.

Fixes #8066

~~_This is what happens when you force the riff-raff to learn a little scheme._~~

ihnorton committed:
Implement __LINE__ magic macros

Fixes #8066

ihnorton commented:
Alternatively, could use the same pattern just for `__LINE__` (and also for `__FILE__` to make the implementation a little bit [more robust](https://github.com/JuliaLang/julia/issues/8066#issuecomment-128103235)?)

timholy commented:
> This is what happens when you force the riff-raff to learn a little scheme.

Quote of the week!

ScottPJones commented:
:clap::clap::clap:! :+1:

StefanKarpinski commented:
Nice. Probably needs a once over by @JeffBezanson

JeffBezanson commented:
A `LineNumberNode` used as data should probably be wrapped in `quote`, since bits of the compiler might assume that LineNumberNodes don't execute, or can always be removed safely.

Wouldn't it be simpler just to replace `@__LINE__` with the result of `input-port-line` at parse time?

ihnorton commented:
That is much nicer. Updated to do so for both `@__LINE__` and `@__FILE__`.

(I'm not sure how to document these now -- writing doc strings lead to a "Invalid doc expression" during sysimg build. Removed for now, will open a separate issue if I can't figure it out)

JeffBezanson commented on src/julia-parser.scm:
```diff
@@ -2039,14 +2039,18 @@
-              (if (ts:space? s)
+              (cond
+                 ((eqv? head '__LINE__) (input-port-line (ts:port s)))
+                 ((eqv? head '__FILE__) `(call string (quote ,current-filename)))
```
Could just be a constant string, `(string current-filename)`.

JeffBezanson commented:
I'm not 100% sure how to make this doc-able either. Maybe we should parse `@__LINE__` as `(macrocall @__LINE__ 42)`. Then we could define it as an identity macro and document it there.

mbauman commented:
This is awesome.  You could document them as [keywords](https://github.com/JuliaLang/julia/blob/master/base/docs/basedocs.jl#L21), which are checked first:

```jl
julia> Base.Docs.keywords[symbol("@__FILE__")] = "filedoc"
"filedoc"

help?> @__FILE__
search: @__FILE__

"filedoc"
```

ihnorton commented:
Perfect, thanks @mbauman. Updated.

Implementing `@__FILE__` this way changes the result when used in `include_string(..., fake_filename)` ... It seems unlikely that anyone relies on this behavior, but I will revert that change if there are any objections.

JeffBezanson commented:
A filename gets passed through from `include_string` to `jl-parse-string-stream`, so I think it's possible to still get the right filename. Is that not happening?

stevengj commented:
@ihnorton, IJulia relies on the `@__FILE__` behavior for `include_string(..., fake_filename)`, where `fake_filename` is e.g. `In[23]`, in order to get backtraces correct.

stevengj commented on test/loading.jl:
```diff
@@ -3,8 +3,10 @@
+thefname = "the fname!//\\&\1*"
+# passing a second argument to include_string should be reflected by @__FILE__
+@test include_string("test_atsign_file() = @__FILE__", thefname)() == thefname
+@test include_string("test_atsign_file() = @__FILE__")() == "string"
```
No test for `@__LINE__`?

ihnorton commented:
Sorry, I misunderstood what was going on with the test failure. This does not change the behavior of `include_string` at all, but rather of `@__FILE__`. `include_string(..., filename)` still works fine. The change is that now `@__FILE__` actually reflects the filename that the parser sees (such as the argument to `include_string`). This feels arguably more accurate.

with `/tmp/foo.jl` as:
```
println(include_string("@__FILE__", "foo"))
println(Base.source_path())
```

on master the out of `julia /tmp/foo.jl` is:
```
/tmp/foo.jl
/tmp/foo.jl
```

on this branch:
```
foo
/tmp/foo.jl
```

There are only two packages that use `include_string` -- IJulia and Jewel -- and their usage should still work fine. On the other hand, a large number of packages use `@__FILE__`. I don't know of any other code path whereby `@__FILE__` would not equal `Base.source_path`, but it is possible that this change could break some code.

ihnorton commented:
I removed the `@__FILE__` change for now. Will merge soon if there are no objections.

ihnorton commented:
... or not, I guess. Just saw the feature freeze announcement.

tkelman commented on base/docs/helpdb.jl:
```diff
@@ -9537,12 +9539,20 @@ Returns an iterator over substrings of `s` that correspond to the extended graph
 
-`@__FILE__` expands to a string with the absolute path and file name of the script being run. Returns `nothing` if run from a REPL or an empty string if evaluated by `julia -e <expr>`.
+`@__FILE__` expands to a string with the absolute path and file name of the script being run.
+Returns `"none"` if run from a REPL or command-line context.
```
Did this actually change from `nothing` to `"none"` here?

tkelman replied:
@ihnorton can you confirm? Just tried it and this looks wrong.

ihnorton replied:
Yeah that was left from the proposed change. I'll fix the doc.
On Aug 28, 2015 7:35 PM, "Tony Kelman" <notifications@github.com> wrote:

> In base/docs/helpdb.jl
> <https://github.com/JuliaLang/julia/pull/12727#discussion_r38254310>:
>
> >      @__FILE__() -> AbstractString
> >
> > -`@__FILE__` expands to a string with the absolute path and file name of the script being run. Returns `nothing` if run from a REPL or an empty string if evaluated by `julia -e <expr>`.
> > +`@__FILE__` expands to a string with the absolute path and file name of the script being run.
> > +Returns `"none"` if run from a REPL or command-line context.
>
> @ihnorton <https://github.com/ihnorton> can you confirm? Just tried it
> and this looks wrong.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/12727/files#r38254310>.
>

ihnorton replied:
8fb70464dfb16dd8b908d562e63f95eb1845d589 (on the docfix branch)

StefanKarpinski commented:
Let's just pretend this got merged before that announcement.
